MON-3669: Add rule for effective CPU cores for subscription usage purposes

kahowell opened the pull request:
Red Hat OpenShift Container Platform subscriptions are often measured against underlying cores. However, the metrics for cores are unreliable with some known edge cases. Namely, when virtualization is used, depending on a variety of factors, the hypervisor doesn't report the underlying cores, and instead reports a core per "cpu" where "cpu" is a schedulable executor (possibly backed by a single hyperthreaded executor). In order to address, we assume a ratio of 2-vCPU to 1 core, and divide the "cores" value by 2 to normalize when we detect that hyperthreading information was not reported, when we're on x86-64 CPU architecture.

At this time, x86-64 clusters with HT undetected or turned off are the ones affected.

The rule implementation itself divides the cluster into the following components:

* worker x86-64 nodes having hyperthreading information reported (no adjustment)
* worker x86-64 nodes having no hyperthreading information reported (divide by 2)
* worker nodes with non-x86-64 CPU arch (no adjustment)
* schedulable control plane x86-64 nodes having hyperthreading information reported (no adjustment)
* schedulable control plane x86-64 nodes having no hyperthreading information reported (divide by 2)
* schedulable control plane nodes with non-x86-64 CPU arch (no adjustment)

simonpasquier reviewed (commented):
I see no reason why the rule shouldn't be deployed on each cluster by the cluster monitoring operator.

moadz reviewed (commented):
So running the query in info-gw returns the following error [link](https://infogw-data.api.openshift.com/graph?g0.expr=%20%23%20worker%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20virt%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20((sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)%20*%20on(_id)%20cluster_infrastructure_provider%7Btype!%3D%22BareMetal%22%7D)%20%2F%202.0%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20phys%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20cluster_infrastructure_provider%7Btype%3D%22BareMetal%22%7D%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20virt%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20((sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20*%20on(_id)%20cluster_infrastructure_provider%7Btype!%3D%22BareMetal%22%7D)%20%2F%202.0%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20phys%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)*%20on(_id)%20cluster_infrastructure_provider%7Btype%3D%22BareMetal%22%7D%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))&g0.tab=1&g0.stacked=0&g0.range_input=1h&g0.max_source_resolution=0s&g0.deduplicate=1&g0.partial_response=0&g0.store_matches=%5B%5D&g0.engine=prometheus&g0.explain=0):

```error
many-to-many matching not allowed: matching labels must be unique on one side
```

I believe the problematic query is [this](https://infogw-data.api.openshift.com/graph?g0.expr=((sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)%20*%20on(_id)%20cluster_infrastructure_provider%7Btype!%3D%22BareMetal%22%7D)%20%2F%202.0%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))&g0.tab=1&g0.stacked=0&g0.range_input=1h&g0.max_source_resolution=0s&g0.deduplicate=1&g0.partial_response=0&g0.store_matches=%5B%5D&g0.engine=prometheus&g0.explain=0) one.

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,55 @@
+                +
+
+                # schedulable control plane ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="true"})* on(_id) max by (_id) (cluster_master_schedulable) or (0 * max by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum)))
```
unless I'm mistaken, it sums cores for all nodes while it should filter master nodes?

moadz replied:
Not certain... if we pick a cluster and do it with/without is it what we expect minus the control plane?

kahowell replied:
Correct, I've been using a [python script](https://gist.github.com/kahowell/56368cd6aac4e3d2bb1176266c9d02a5) to generate the query for consistency, and it had a (now-fixed) typo in it on line 14 :disappointed:.

moadz commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,55 @@
+                +
+
+                # worker phys non-ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra!="true",label_node_role_kubernetes_io_master!="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"})* on(_id) cluster_infrastructure_provider{type="BareMetal"} or (0 * max by (_id) (node_role_os_version_machine:cpu_c ...
```
```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra!="true",label_node_role_kubernetes_io_master!="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"})* on(_id) group(cluster_infrastructure_provider{type="BareMetal"}) by (_id) or (0 * max by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum)))
```

moadz commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,55 @@
+                +
+
+                # schedulable control plane phys non-ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"})* on(_id) max by (_id) (cluster_master_schedulable)* on(_id) cluster_infrastructure_provider{type="BareMetal"} or (0 * max by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum)))
```
```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"})* on(_id) max by (_id) (cluster_master_schedulable)* on(_id) group(cluster_infrastructure_provider{type="BareMetal"} ) by (_id) or (0 * max by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum)))
```

moadz reviewed (commented):
Group by worked, but some of the subqueries need the same fix. 

On the amended [query](https://infogw-data.api.openshift.com/graph?g0.expr=%23%20worker%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20virt%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20((sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)%20*%20on(_id)%20group(cluster_infrastructure_provider%7Btype!%3D%22BareMetal%22%7D)%20by%20(_id))%20%2F%202.0%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20phys%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20group(cluster_infrastructure_provider%7Btype%3D%22BareMetal%22%7D)%20by%20(_id)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20virt%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20((sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20*%20on(_id)%20group(cluster_infrastructure_provider%7Btype!%3D%22BareMetal%22%7D)%20by%20(_id))%20%2F%202.0%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20phys%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)*%20on(_id)%20group(cluster_infrastructure_provider%7Btype%3D%22BareMetal%22%7D)%20by%20(_id)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)*%20on(_id)%20max%20by%20(_id)%20(cluster_master_schedulable)%20or%20(0%20*%20max%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum)))&g0.tab=1&g0.stacked=0&g0.range_input=1h&g0.max_source_resolution=0s&g0.deduplicate=1&g0.partial_response=0&g0.store_matches=%5B%5D&g0.engine=prometheus&g0.explain=0) it took about 11s to evaluate in production (which should be manageable). 

<img width="339" alt="image" src="https://github.com/openshift/telemeter/assets/50833398/0bed9475-295b-40da-b9e6-9ca3c19332fe">

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,56 @@
+                +
+
+                # worker phys non-ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra!="true",label_node_role_kubernetes_io_master!="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"})* on(_id) cluster_infrastructure_provider{type="BareMetal"} or (0 * max by (_id) (node_role_os_version_machine:cpu_c ...
```
I can see clusters where cluster_infrastructure_provider has `type="BareMetal"` but all nodes are reported as vmware in `cluster:virt_platform_nodes:sum`. IIUC the query will not divide node_role_os_version_machine:cpu_capacity_cores:sum by 2 in this case.

kahowell replied:
@simonpasquier oof, that's a gap for sure. I'm going to discuss with stakeholders, in order to address properly we might actually need an in-cluster monitoring change.

kahowell replied:
@simonpasquier we're currently thinking we'll simplify a bit and handle those use cases by not attempting to distinguish physical and virtual nodes. If hyperthreading isn't detected, then we'll divide by 2. This seems reasonable, as OCP subs use as a unit of measure "Cores or vCPUs" with a ratio of 2 vCPUs per core. In practice, a 4 CPU physical node, backed by 4 physical cores, would be counted as "2 cores". (Note if this same node was deployed on the same hardware, but on a VM, it'd also be counted as "2 cores"). If hyperthreading were enabled for that same hardware, it'd be reported as 4 cores, and a VM would have access to 8 vCPUs. So it's at least consistent.

kahowell commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,56 @@
+              record: 'cluster:usage:workload:capacity_effective_cpu_cores',
+              expr: |||
+                # worker ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra!="true",label_node_role_kubernetes_io_master!="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="true"}) or (0 * max by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum)))
```
Implemented, thanks for the suggestion!

kahowell commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,56 @@
+              // 2. x86_64 nodes that are do not show hyperthreading and are bare metal have an accurate cores value in node_role_os_version_machine:cpu_capacity_cores:sum.
+              // 3. x86_64 nodes that are do not show hyperthreading and are not bare metal need the cores value adjusted to account for 2 threads per core (* 0.5).
+              // 4. Other CPU architectures are assumed to have accurate values in node_role_os_version_machine:cpu_capacity_cores:sum.
+              record: 'cluster:usage:workload:capacity_effective_cpu_cores',
```
Name changed to cluster:capacity_effective_cpu_cores.

kahowell commented:
@moadz @simonpasquier I've addressed PR feedback, and I also simplified a bit, see https://github.com/openshift/telemeter/pull/501#discussion_r1462481721 for rationale around some of the simplification. I updated the PR description accordingly.

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,51 @@
+              // returns 0 for any cluster reporting core capacity, used to improve performance of cluster:capacity_effective_cpu_cores
+              record: 'cluster:cpu_capacity_cores:_id',
+              expr: |||
+                group by(_id) (node_role_os_version_machine:cpu_capacity_cores:sum) * 0
```
(nit) the next rule only looks at rhcos nodes, we can filter away clusters without any rhcos nodes.
```suggestion
                group by(_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos"}) * 0
```

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,51 @@
+                +
+
+                # schedulable control plane ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="true"}) * on(_id) group by(_id) (cluster_master_schedulable) or cluster:cpu_capacity_cores:_id)
```
the current expression counts all cores of the master nodes even when cluster_master_schedulable is 0.

you need either (my preference)

```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="true"}) * on(_id) group by(_id) (cluster_master_schedulable == 1) or cluster:cpu_capacity_cores:_id)
```
or 

```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="true"}) * on(_id) max by(_id) (cluster_master_schedulable) or cluster:cpu_capacity_cores:_id)
```

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,51 @@
+                +
+
+                # schedulable control plane non-ht amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"}) * on(_id) group by(_id) (cluster_master_schedulable) / 2.0 or cluster:cpu_capacity_cores:_id)
```
```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch="amd64",label_node_hyperthread_enabled="false"}) * on(_id) group by(_id) (cluster_master_schedulable == 1) / 2.0 or cluster:cpu_capacity_cores:_id)
```

simonpasquier commented on jsonnet/telemeter/rules.libsonnet:
```diff
@@ -160,6 +160,51 @@
+                +
+
+                # schedulable control plane non-amd64
+                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch!="amd64"}) * on(_id) group by(_id) (cluster_master_schedulable) or cluster:cpu_capacity_cores:_id)
```
```suggestion
                (sum by (_id) (node_role_os_version_machine:cpu_capacity_cores:sum{label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",label_kubernetes_io_arch!="amd64"}) * on(_id) group by(_id) (cluster_master_schedulable == 1) or cluster:cpu_capacity_cores:_id)
```

simonpasquier commented:
Eventually we need to add unit tests for this new rule (and all existing rules) but there's no reason to block the PR on this. I'll create a follow-up ticket.

simonpasquier commented:
/ok-to-test

simonpasquier commented:
/retitle MON-3669: Add rule for effective CPU cores for subscription usage purposes

simonpasquier commented:
/approve

I'll let the final review to @moadz

moadz commented:
/lgtm 
![image](https://github.com/openshift/telemeter/assets/50833398/f1c2b469-dc20-4171-a056-5773bbe3f7d1)
Performance looks good, current interval is 4m, and the rule group takes about 120s to evaluate: 
![image](https://github.com/openshift/telemeter/assets/50833398/52b474c0-12b6-4c52-96f4-7697d188411c)

Checked the expanded query [here](https://infogw-proxy.api.openshift.com/graph?g0.expr=%20%23%20worker%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)%20%2F%202.0%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20worker%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_infra!%3D%22true%22%2Clabel_node_role_kubernetes_io_master!%3D%22true%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_master%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22true%22%7D)%20*%20on(_id)%20group%20by(_id)%20(cluster_master_schedulable%20%3D%3D%201)%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20non-ht%20amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_master%3D%22true%22%2Clabel_kubernetes_io_arch%3D%22amd64%22%2Clabel_node_hyperthread_enabled%3D%22false%22%7D)%20*%20on(_id)%20group%20by(_id)%20(cluster_master_schedulable%20%3D%3D%201)%20%2F%202.0%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20schedulable%20control%20plane%20non-amd64%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(sum%20by%20(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%2Clabel_node_role_kubernetes_io_master%3D%22true%22%2Clabel_kubernetes_io_arch!%3D%22amd64%22%7D)%20*%20on(_id)%20group%20by(_id)%20(cluster_master_schedulable%20%3D%3D%201)%20or%20group%20by(_id)%20(node_role_os_version_machine%3Acpu_capacity_cores%3Asum%7Blabel_node_openshift_io_os_id%3D%22rhcos%22%7D)%20*%200)&g0.tab=1&g0.stacked=0&g0.range_input=8h&g0.max_source_resolution=auto&g0.deduplicate=1&g0.partial_response=0&g0.store_matches=%5B%5D&g0.engine=prometheus&g0.explain=0&g0.step_input=3600).

moadz commented:
/approve
